apps/app_utf8.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <openssl/buffer.h>
+#include <openssl/asn1.h>        /* For UTF8_putc */
+#include "apps.h"
+
+char *to_utf8(const char *str)
+{
+    char *str_utf8 = NULL;
+
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199409L \
+    || defined(_WIN32)
+    wchar_t *wstr = NULL;
+    size_t wstr_size;
+    size_t utf8_count, i, j;
+
+    /*
+     * Because MultiByteToWideChar returns 0 for error,
+     * avoid trying to convert empty strings
+     */
+    if (str[0] == '\0')
+        return OPENSSL_strdup("");
+
+# ifdef _WIN32
+    if ((wstr_size = MultiByteToWideChar(CP_ACP,
+                                         MB_ERR_INVALID_CHARS | MB_PRECOMPOSED,
+                                         str, -1, NULL, 0)) == 0
+        || (wstr = OPENSSL_malloc(wstr_size * sizeof(*wstr))) == NULL)
+        return NULL;
+    MultiByteToWideChar(CP_ACP, MB_ERR_INVALID_CHARS | MB_PRECOMPOSED,
+                        str, -1, wstr, wstr_size);
```

Comment:
This should be conditional on presence of OPENSSL_WIN32_UTF8 environment variable. Indeed, if it's set, then str is already UTF-8, and re-interpreting it as CP_ACP would ruin it. This is unlike mbstowcs. I mean if locale is UTF-8, then mbstowcs yields equivalent string. "Equivalent" means that conversion with UTF8_putc would yield original string.

Reply:
In case question pops up if in such case one can use mbstowcs even Windows. No. setllocale(LC_CTYPE,"") sets non-UTF8 locale, and so mbstowcs would be equivalent to MultiByteToWideChar with CP_ACP, which will ruin UTF-8 string returned with OPENSSL_WIN32_UTF8.

Reply:
> setllocale(LC_CTYPE,"") sets non-UTF8 locale [on Windows]

Modulo fact that I'm not sure how things are on Asian Windows version...

Reply:
Right, forgot about that env variable for a moment.  Do I have to use `GetEnvironmentVariableW` or does `getenv` suffice?  I mean, Considering we're not checking the value anyway, not its value...

Reply:
You don't *have to* use GetEnvironmentVariableW, but doing so is stylish. Why is that? Do you know how getenv works on Windows? The system interface is all Unicode, so that getenv would have to convert argument to Unicode (with MultiByteToWideChar to dynamically allocated buffer) make the call, find '=', convert value from Unicode (with WideCharToMultiByte)... In other words it's somewhat wasteful... Well, not that we have to care that much about that kind of waste of resources, but even besides that, using actual Win32 interface in #ifdef _WIN32 segment is only appropriate and is a nice touch :-)

Reply:
Okie

> using actual Win32 interface in #ifdef _WIN32 segment is only appropriate and is a nice touch :-)

I can understand the feeling although I don't share it.  For VMS, the only reason I'd see to use VMS specific functions is when they do a better enough job than the standard C or POSIX functions.
But no matter, I can certainly adapt to your sense of elegance :-)

Reply:
Done

Reply:
But `OPENSSL_WIN32_UTF8` only affects the default console I/O PEM callback, doesn't it? It doesn't change behaviour of command line options or strings obtained from GUI code? Those are still not going to be UTF-8.